src/initilization-context.ts

```diff
@@ -18,8 +18,12 @@ export class InitializationContext {
 		this.tasks.add(task);
 		task.then(() => {
 			this.tasks.delete(task);
-			if (this.tasks.size === 0)
-				this.waiting.forEach(done => done());
+			// allow additional tasks to be queued as a result of this one
+			Promise.resolve().then(() => {
```

Comment:
Wait on an immediately resolved promise to make sure we check the `tasks` list _after_ allowing other tasks to be queued.

Reply:
I understand in the abstract sense that async init code could queue other async init code, but its difficult to make out the precise scenario just looking at the code. Just to make sure I understand... A `wait()` task was queued to async resolve a value via `tryResolveValue` (ex: a lookup entity), and when it completed we would fire off the queue callbacks (the `waiting` list) - which would resolve the context created for "create" calls and/or constructor calls. But, if the queued task (i.e. the `tryResolveValue` call) required additional waits on the context due to nested calls to constructors or "create" (via a "set" call after value resolution?), then we want it to wait for those to get queued up (the context's promise callback may fire before the value resolver makes a recursive call to `wait`?), hence the immediately resolving promise.

Reply:
Also, its a minor thing, but I think the `done` naming used in the previous version made it a bit more clear. And the use of `shift()()` is harder to grok.

Reply:
Also, just a thought: what do you think about a new helper method for this (ex: `nextTick`)

Reply:
Your understanding is correct. We shouldn't try to run the `ready()` callbacks until the caller of `tryResolveValue` has a chance to process the result.

I will update the `shift()()` code to be more clear.

Reply:
Updated